packages/runtime/container-runtime/src/pendingStateManager.ts

```diff
@@ -74,6 +74,23 @@ function buildPendingMessageContent(
 	return JSON.stringify({ type, contents, compatDetails });
 }
 
+class OrderedDeque extends Deque<IPendingMessage> {
+	public push(...items: IPendingMessage[]): number {
+		// We need to ensure that the items are pushed in order of their reference sequence number.
+		let totalPushed = this.length;
+		for (const item of items) {
+			const lastItem = this.peekBack();
+			assert(
+				lastItem === undefined ||
+					item.referenceSequenceNumber >= lastItem.referenceSequenceNumber,
+				"Items must be pushed in order of their reference sequence number",
```

Comment:
To be honest, I'm not sure why this is needed, or what it does.
I think all it does is testing that ops will be added to pending list in non-decreasing reference sequence numbers. I think it's fine to just assume that's the case (without extra checking), as Container's reference sequence number (or same as last op's processed sequence number) can't move backwards - DeltaManager will for sure validate that.
There could be bugs in pendingStateManager, of course, but this is just too fundamental assumption that will break everything if it's gone wrong (even if it's at the scale of pending state manager).

If you really want to assert that, I'd rather assert that at callsite where messages are added to pending queue. No need to test initialMessages as messages move to it from pending queue, so they have same order there as they had in pending queue (plus see my remark about Daniel's PR below)

Reply:
Removed